lib/ui/plugins.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of dart.ui;
+
+/// Functionality for Flutter plugin authors.
+abstract class PluginUtilities {
```

Comment:
These are needed to lookup callback methods which are top-level functions or static methods in a class by name. The general flow is something like this:

- Plugin user passes callback `fooBar` to the plugin.
- The plugin calls `PluginUtilities.getPathForFunctionLibrary`, `PluginUtilities.getNameOfFunction`, and `PluginUtilities.getNameOfFunctionClass` to get the library URL, function name, and class name (if applicable). These values are passed from native as strings to a callback handler (this is a Dart method) setup by the plugin  when a background event is triggered (e.g., `AlarmManager` fires).
- When in the callback handler is invoked, `PluginUtilities.getClosureByName` is called with the function name, class name, and library path to retrieve the closure for the plugin user's callback. The plugin can then invoke said closure.

Reply:
I am concerned about this API. From most serious to least serious concerns:

* If someone uses this API directly with input that is in any way under the control of the user, it's a ripe for becoming an attack vector, where the user causes the app to jump to arbitrary top-level static functions.

* This seems like it will be abused. For example, someone might store this data, pass it over the network to another instance of the program, then look up the address to jump to and do that. This would be very sketchy code, but this API doesn't make it feel bad.

* What happens if someone caches these values across version updates, then uses them in a subsequent version of their code where the function has been tree-shaken out?

* Least serious: if we have to have this API, it seems like we should have a single type that represents the path, name, and class of the closure.

Can you talk more broadly about what problem these methods solve? Maybe we can find a safer solution to this problem.

Reply:
Good point. I feel like a lot of your concerns could be addressed by an opaque class with static methods that could:
* Create an instance of the class given a closure which contains all the path, name, and class information.
* Take in an instance of this class and return the closure it references.

This way, we can avoid having these strings visible and modifiable by the user, prevent abuse of what are basically Dart VM APIs, and make caching difficult. The only difficulty I see with this approach is passing said object over method channels since it would have to be serialized and deserialized in such a way that the data itself still remains private, but anyone can listen to a method channel if they know the name of the channel. I'd have to give this part some thought...

Reply:
Your latest update resolves my concerns. Thanks!